### PR TITLE
Lint + Format Megablocks

### PR DESCRIPTION
**Pull request by eitanturok**
This PR formats and lints all the code in MegaBlocks. It also adds a pre-commit-config.yaml file to do all of this. This is the first step of creating CI/CD for MegaBlocks.

These changes primarily consist of:
1. removing unnecessary semicolons, extra spaces, and package imports
2. replacing lambdas with actual functions, i.e. def f():
3. replacing double quotes with single quotes
4. Adding a license to the beginning of every file

Many of these changes were generated automatically by running the `pre-commit-config.yaml` file which contains `astral-sh/ruff-pre-commit`, `google/yapf`, `pycqa/isort`, `adrienverge/yamllint`, `pre-commit/pre-commit-hooks`, `Lucas-C/pre-commit-hooks`, `kynan/nbstripout`, `trufflesecurity/trufflehog`. This `pre-commit-config.yaml` was taken from `composer`'s `pre-commit-config.yaml`.

Initially, I was planning on making these all separate, smaller PR's but @b-chu said I should just make this one single PR. I am currently running a small MoE training script with MCLI to confirm that this does not break anything.

To do:
1. Should I delete the unused variables? I currently just commented them out. You can see the exact variables I commented out in this commit: [comment out unused variables; delete later](https://github.com/databricks/megablocks/commit/b8d4eb70cf5f0f4dc26a50ac3bddeac694e0f3d4).
2. I made the license 
> Copyright 2024 MosaicML MegaBlocks authors
> SPDX-License-Identifier: Apache-2.0

Is license correct from a legal perspective? Do we want to include anything about giving Trevor credit?
3. It seems like running the Ruff formatter deleted the imports in every __init__.py file. I'm not sure if this is desirable behavior. Should I revert this?

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+#   hooks:
+#   - id: pydocstyle
+#     name: pydocstyle
+#     entry: pydocstyle
+#     language: python
+#     types: [python]
+#     exclude: (.ci|.github)
+#     additional_dependencies:
+#     - "toml"
+#   rev: 6.1.1
```
Comment (b-chu) [inline .pre-commit-config.yaml:119]: Why is this commented out?

File: .yamllint.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+
+rules:
+  line-length:
+    max: 1000 # effectively we do not enforce line length
```
Comment (b-chu) [inline .yamllint.yaml:5]: This should follow Composer

File: LICENSE [hunk 3]
```diff
@@ -387,4 +387,4 @@
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
-   limitations under the License.
\ No newline at end of file
+   limitations under the License.
```
Comment (b-chu) [inline LICENSE:390]: What's the change here?

File: docker.sh
Comment (b-chu) [inline docker.sh:1]: Why are there all these empty file changes?

File: megablocks/__init__.py [hunk 4]
```diff
@@ -1,2 +1,2 @@
-import megablocks.layers.dmoe
-import megablocks.layers.moe
```
Comment (b-chu) [inline megablocks/__init__.py:2]: Why is this deleted?

File: megablocks/backend/kernels.py [hunk 5]
```diff
@@ -39,19 +43,10 @@ def assert_equal(a, b):
...
-        padded_bins,
-        NUM_COLUMNS : tl.constexpr,
-        TOP_K : tl.constexpr,
-        BLOCK_X : tl.constexpr,
-        A_TO_B : tl.constexpr,
-        SCALE : tl.constexpr):
+def _padded_copy(a, b, indices, bin_ids, weights, bins, padded_bins,
+                 NUM_COLUMNS: tl.constexpr, TOP_K: tl.constexpr,
+                 BLOCK_X: tl.constexpr, A_TO_B: tl.constexpr,
+                 SCALE: tl.constexpr):
```
Comment (b-chu) [inline megablocks/backend/kernels.py:49]: This should have trailing comma

File: megablocks/backend/kernels.py [hunk 6]
```diff
@@ -262,8 +243,8 @@ def _padded_copy_wgrad(
     x += tl.multiple_of(index_x * NUM_COLUMNS, NUM_COLUMNS)
     offsets = tl.max_contiguous(tl.arange(0, BLOCK_X), BLOCK_X)
 
-    acc = tl.zeros((BLOCK_X,), dtype=tl.float32)
-    iterations = tl.cdiv(NUM_COLUMNS, BLOCK_X)
+    acc = tl.zeros((BLOCK_X, ), dtype=tl.float32)
+    # iterations = tl.cdiv(NUM_COLUMNS, BLOCK_X)
```
Comment (b-chu) [inline megablocks/backend/kernels.py:247]: Why is this commented out?

**Review by b-chu — CHANGES_REQUESTED** [review]
Left some initial comments. Can you first make a PR that adds GH CI/CD? Then for this PR you also need a pyproject.toml.

File: docker.sh
Comment (eitanturok) [reply docker.sh:1]: these files are not empty. if you click on it you can see that there is still code in them. 

Not sure why they are being marked as empty.

This holds true for all the `.sh` files.

File: megablocks/backend/kernels.py (hunk 6 above)
Comment (eitanturok) [reply megablocks/backend/kernels.py:247]: This was an unused parameter and rather than deleting it, I just commented out. I flagged this above.

I think I'm going to delete all of these unused parameters and if things break, I'll put them back.

**Comment by eitanturok** [comment]
I renamed this branch to dev and this automatically closed this PR. I will reopen this in a different PR.

**Comment by eitanturok** [comment]
Continued this work in #123 .
